src/ellswift.rs

```diff
@@ -0,0 +1,605 @@
+    /// # Example
+    /// ```
+    /// use secp256k1::{ellswift::ElligatorSwift, PublicKey, Secp256k1, SecretKey};
+    /// let secp = Secp256k1::new();
+    /// let sk = SecretKey::from_slice(&[1; 32]).unwrap();
+    /// let pk = PublicKey::from_secret_key(&secp, &sk);
+    /// let es = ElligatorSwift::from_pubkey(pk);
+    /// ```
+    pub fn from_pubkey(pk: PublicKey) -> ElligatorSwift {
+        Self::ellswift_encode(pk)
+    }
+    /// Computes a shared secret only known by Alice and Bob. This is obtained by computing
+    /// the x-only Elliptic Curve Diffie-Hellman (ECDH) shared secret between Alice and Bob.
+    /// # Example
+    /// ```
+    /// use secp256k1::{
+    ///     ellswift::{ElligatorSwift, ElligatorSwiftParty},
+    ///     PublicKey, SecretKey, XOnlyPublicKey, Secp256k1,
+    /// };
+    /// use core::str::FromStr;
+    ///
+    /// let secp = Secp256k1::new();
+    ///
+    /// let alice_sk = SecretKey::from_str("e714e76bdd67ad9f495683c37934148f4efc25ce3f01652c8a906498339e1f3a").unwrap();
+    /// let bob_sk = SecretKey::from_str("b6c4b0e2f8c4359caf356a618cd1649d18790a1d67f7c2d1e4760e04c785db4f").unwrap();
+    ///
+    /// let alice_es = ElligatorSwift::from_seckey(&secp, alice_sk, None);
+    /// let bob_es = ElligatorSwift::from_seckey(&secp, bob_sk, None);
+    ///
+    /// let alice_shared_secret = ElligatorSwift::shared_secret(alice_es, bob_es, alice_sk, ElligatorSwiftParty::A, None);
+    /// let bob_shared_secret = ElligatorSwift::shared_secret(alice_es, bob_es, bob_sk, ElligatorSwiftParty::B, None);
+    ///
+    /// assert_eq!(alice_shared_secret, bob_shared_secret);
+    /// ```
+    pub fn shared_secret(
+        ellswift_a: ElligatorSwift,
+        ellswift_b: ElligatorSwift,
+        secret_key: SecretKey,
+        party: ElligatorSwiftParty,
+        data: Option<Vec<u8>>,
+    ) -> ElligatorSwiftSharedSecret {
+        Self::shared_secret_with_hasher(ellswift_a, ellswift_b, secret_key, party, data, None)
+    }
+    /// Constructs an ElligatorSwift object from a 64-byte hex string.
+    /// # Example
+    /// ```
+    /// use secp256k1::ellswift::ElligatorSwift;
+    /// let es = ElligatorSwift::from_hex("c01204f76d2f886a4d3319e2ac0c86a2cb560ec29c732336a99ff152");
+    /// ```
+    pub fn from_hex(hex: &str) -> ElligatorSwift {
+        let mut ser = [0u8; 64];
+        from_hex(hex, &mut ser).unwrap();
+        ElligatorSwift::from_array(ser)
+    }
+    /// Computes a shared secret, just like `shared_secret`, but with a custom hash function
+    /// for computing the shared secret. For compatibility with other libraries, you should
+    /// use `shared_secret` instead, which is already compatible with BIP324.
+    pub fn shared_secret_with_hasher(
+        ellswift_a: ElligatorSwift,
+        ellswift_b: ElligatorSwift,
+        secret_key: SecretKey,
+        party: ElligatorSwiftParty,
+        data: Option<Vec<u8>>,
+        hasher: EllswiftECDHHashFn,
```

Comment:
Feel free to look here for a rustier API: https://github.com/rust-bitcoin/rust-secp256k1/commit/af8fa21a20d7d747a9bafcd8c4f1f3fedb51675e